### PR TITLE
Move React components to separate package and unify data handling

### PR DESCRIPTION
**Pull request by martin-fleck-at**
- Extract React components to separate, Theia-less package
- Provide shared layer (reducer, context, provider) for all views
- Align usage fo ErrorView to display errors
- Improve overall typing of model reducer functionality (actions)
- Use 'immer' hooks to ease changes to immutable state handling

### CODE REVIEW

**Review by harmen-xb — COMMENTED** [review]
@martin-fleck-at 

I pushed the changes we talked about in a new commit. I think this refactoring is enough for now to at least make the views common ones which are shared across property and form widgets.

As seen in 1 review comment, I think we can still remove some dependencies in the form-client (since we don't use the UI stuff there anymore).

File: packages/form-client/style/form-view.css
Comment (martin-fleck-at) [inline packages/form-client/style/form-view.css:34]: Is there a particular reason you left `.form-editor-attributes-tab` and `.react-tabs__tab > h3` in this file? I believe `.form-editor-attributes-tab` could probably be removed as it is not used anywhere (anymore) and the other one should probably also go into the React package.

In general, we should think about styling a bit more as currently the UI only looks good in 'light' theme but switching to dark theme some parts become unreadable. This should be done in a separate issue, of course.

File: packages/react-model-ui/src/views/common/EntityAttributesDataGrid.tsx
Comment (martin-fleck-at) [inline packages/react-model-ui/src/views/common/EntityAttributesDataGrid.tsx:42]: In the previous property view attributes grid, there was a validation that checks if a name is empty and if it is, it sets that row as erroneous by applying a different CSS class (`entity-attribute-error-row`). Has this been removed on purpose? I think it is fine since we really didn't do any particular client-side validation so we should properly think about it but I just wanted to point that out.

File: packages/react-model-ui/src/ModelReducer.tsx
Comment (martin-fleck-at) [inline packages/react-model-ui/src/ModelReducer.tsx:31]: It seems that after adding the more generic `'entity:attribute:update` action, the actions for name and description are no longer necessary, only the datatype one and even that one could also be mapped to the generic action, what do you think?

**Review by martin-fleck-at — COMMENTED** [review]
Functionally, everything looks good to me. However, I do have a few comments on the commits you made.

Also, since I opened the PR, I cannot approve it, so once you are happy, please approve it and we can merge it.

File: packages/form-client/package.json
Comment (martin-fleck-at) [reply packages/form-client/package.json:39]: Yes, we can, good catch! Also in the property-view.

File: packages/property-view/src/browser/model-property-widget.tsx
Comment (harmen-xb) [reply packages/property-view/src/browser/model-property-widget.tsx:27]: I added it because it's also in the form-editor-widget: https://github.com/CrossBreezeNL/crossmodel/blob/059aee225e24bf1e18f240f7a71163d2124b3a03/packages/form-client/src/browser/form-editor-widget.tsx#L46C1-L47C1

Can it be removed in both places then?

File: packages/react-model-ui/src/views/common/EntityAttributesDataGrid.tsx
Comment (harmen-xb) [reply packages/react-model-ui/src/views/common/EntityAttributesDataGrid.tsx:42]: I indeed purposely remoded it, because it was some custom code which required new state to be maneged. While when I was reading the mui-datagrid docs I saw there is a preprocessing hook for every field.

So I implemented one for the name here: https://github.com/CrossBreezeNL/crossmodel/pull/32/commits/88440351fd9b915bb4e5a9b3ea13e2c3eec4b428#diff-db8e8a38db340fd118288147c3126f021b6261c77da14bfa54d34fa540cfd304R135.

It now shows a tooltip on the field with the error, the updatedRow event is fired.

File: packages/react-model-ui/src/ModelReducer.tsx
Comment (harmen-xb) [reply packages/react-model-ui/src/ModelReducer.tsx:31]: We can indeed remove some at some point.

After yesterdays chat I investigated the DataGrid stuff a bit more about whether the received updatedRow object could be spread (I think thats the right term) into the EntityAttribute object. I think this can be done, but only after we refactor the langium language to have the proper id property and name stuff. In the end we need some kind of row identifier which is numeric in the datagrid. And also the datatype drop-down logic seems a bit unnatural to me. Would be nice if it would just be a cell edit operation just like the other fields where the processRowUpdate thing would process the result.

So based on the above thinking I thought, maybe first finalize this PR. Let you work on upgrading all deps, then refactor the langium grammer to have the proper names in the ast nodes (so no name_val anymore). And then refactor this grid implementation a bit.

In summary: I agree but I would like to prospone that till later.

File: packages/form-client/package.json
Comment (harmen-xb) [reply packages/form-client/package.json:39]: Ok, I will pick this up later

File: packages/form-client/style/form-view.css
Comment (harmen-xb) [reply packages/form-client/style/form-view.css:34]: No particular reason, I just moved the css files. I will clean it up, thanks!

Comment (harmen-xb) [reply packages/form-client/style/form-view.css:34]: In the last checkin the css files are moved and removed where needed.

File: packages/property-view/src/browser/model-property-widget.tsx
Comment (harmen-xb) [reply packages/property-view/src/browser/model-property-widget.tsx]: In the last commit I removed all unneeded undefined assignments, since undefined is always implicit (based on understanding of https://www.tektutorialshub.com/typescript/null-undefined-in-typescript/).

File: packages/form-client/package.json
Comment (harmen-xb) [reply packages/form-client/package.json:38]: @martin-fleck-at  I am trying to set the right dependencies in the package.json files, but I am strugling to understand the react dependency.

Throughout the project I see 2 different type of React imports:

1. import * as React from '@theia/core/shared/react';
2. import * as React from 'react';

If the first one is used there is a dependency on Theia for this package (which is ok, as long as Theia is used on the packahe context). In the second scenario we use 'react', but it's never listed as a dependency in any package.json file.

Shound'nt we have one or the other everywhere? And whats the preferred option?

Comment (martin-fleck-at) [reply packages/form-client/package.json]: Good catch! So we definitely should add `react` as a dependency to our `package.json` of the react-model-ui package but ideally we align with the version that Theia is providing as well (18.2.0 -- `yarn why react`). Theia is [re-exporting](https://github.com/eclipse-theia/theia/tree/master/packages/core#re-exports-mechanism) some common dependencies to make them available through their `@theia/core` package.

However, we purposefully avoided any Theia dependency in the `react-model-ui` package so that - in theory - it could even be re-used outside of Theia. And there really isn't anything Theia-specific that we need either, so in that package using `import * as React from 'react'` is the better approach, given that we actually define that dependency in the package.json.

File: packages/property-view/src/browser/model-property-widget.tsx
Comment (martin-fleck-at) [reply packages/property-view/src/browser/model-property-widget.tsx]: That is correct, thank you!

**Comment by martin-fleck-at** [comment]
@harmen-xb  As we are already touching the dependencies, I saw that you can check the dependencies for each package using `npx depcheck`. It is static so probably only using the import statements but it can give you a good idea of what might be missing or unused. However, it does not check the scripts that we are using our package.json so we should always manually double check if something is truly unused.

**Comment by harmen-xb** [comment]
> o probably only using the import statements but it can give you a good idea of what might be missing or unused. However, it does not check the scripts that we are using our package.json so we should always manually double check if something is truly unused.

I indeed used npx depcheck to see whats needed, thanks!

**Review by harmen-xb — APPROVED** [review]

**Pull request merged**
